pkg/apiserver/resthandler.go

```diff
@@ -59,113 +57,133 @@ type ScopeNamer interface {
+			errorJSON(err, scope.Codec, w)
 			return
 		}
-		ctx := ctxFn(req)
+		ctx := scope.ContextFunc(req)
 		ctx = api.WithNamespace(ctx, namespace)
 
 		result, err := r.Get(ctx, name)
 		if err != nil {
-			errorJSON(err, codec, w)
+			errorJSON(err, scope.Codec, w)
 			return
 		}
-		if err := setSelfLink(result, req, namer); err != nil {
-			errorJSON(err, codec, w)
+		if err := setSelfLink(result, req, scope.Namer); err != nil {
+			errorJSON(err, scope.Codec, w)
 			return
 		}
-		writeJSON(http.StatusOK, codec, result, w)
+		write(http.StatusOK, scope.APIVersion, scope.Codec, result, w, req.Request)
 	}
 }
 
-func parseSelectorQueryParams(query url.Values, version, apiResource string) (label labels.Selector, field fields.Selector, err error) {
-	labelString := query.Get(api.LabelSelectorQueryParam(version))
-	label, err = labels.Parse(labelString)
-	if err != nil {
-		return nil, nil, errors.NewBadRequest(fmt.Sprintf("The 'labels' selector parameter (%s) could not be parsed: %v", labelString, err))
-	}
-
-	convertToInternalVersionFunc := func(label, value string) (newLabel, newValue string, err error) {
-		return api.Scheme.ConvertFieldLabel(version, apiResource, label, value)
-	}
-	fieldString := query.Get(api.FieldSelectorQueryParam(version))
-	field, err = fields.ParseAndTransformSelector(fieldString, convertToInternalVersionFunc)
-	if err != nil {
-		return nil, nil, errors.NewBadRequest(fmt.Sprintf("The 'fields' selector parameter (%s) could not be parsed: %v", fieldString, err))
-	}
-	return label, field, nil
-}
-
 // ListResource returns a function that handles retrieving a list of resources from a rest.Storage object.
-func ListResource(r rest.Lister, ctxFn ContextFunc, namer ScopeNamer, codec runtime.Codec, version, apiResource string) restful.RouteFunction {
+func ListResource(r rest.Lister, rw rest.Watcher, scope RequestScope, forceWatch bool) restful.RouteFunction {
 	return func(req *restful.Request, res *restful.Response) {
 		w := res.ResponseWriter
 
-		namespace, err := namer.Namespace(req)
+		namespace, err := scope.Namer.Namespace(req)
 		if err != nil {
-			errorJSON(err, codec, w)
+			errorJSON(err, scope.Codec, w)
 			return
 		}
-		ctx := ctxFn(req)
+		ctx := scope.ContextFunc(req)
 		ctx = api.WithNamespace(ctx, namespace)
 
-		label, field, err := parseSelectorQueryParams(req.Request.URL.Query(), version, apiResource)
+		out, err := queryToObject(req.Request.URL.Query(), scope, "ListOptions")
 		if err != nil {
-			errorJSON(err, codec, w)
+			errorJSON(err, scope.Codec, w)
+			return
+		}
+		opts := *out.(*api.ListOptions)
+
+		// transform fields
+		fn := func(label, value string) (newLabel, newValue string, err error) {
+			return scope.Convertor.ConvertFieldLabel(scope.APIVersion, scope.Kind, label, value)
+		}
+		if opts.FieldSelector, err = opts.FieldSelector.Transform(fn); err != nil {
+			// TODO: allow bad request to set field causes based on query parameters
+			err = errors.NewBadRequest(err.Error())
```

Comment:
We're working on making all such errors more specific. Is the field selector included in the error message at least?